### PR TITLE
Glowshrooms and their variants no longer spread.

### PR DESCRIPTION
**Pull request by Timberpoes**
## About The Pull Request

Alternative to #56981 
Closes #56981

image
image
image

I got permission from the boss.
image

## Why It's Good For The Game

I no longer have to restart servers and deal with the MC crashing from glowshrooms replicating faster than SDQL2 can delete them.

## Changelog
:cl:
fix: Glowshrooms and their variants no longer spread.
/:cl:

### CODE REVIEW

**Comment by MrDoomBringer** [comment]
about time

**Comment by TemporalOroboros** [comment]
I would rather they be nerfed than removed. What brought this on?

**Comment by Mothblocks** [comment]
did the picture not make it obvious...?

**Review by Mothblocks — APPROVED** [review]

**Comment by Timberpoes** [comment]
> 
> 
> I would rather they be nerfed than removed. What brought this on?

Top screenshot, 10k glowshrooms deleted. They were replicating as fast as SDQL2 could delete them - I ran a query that deleted 12k just before the one screenshotted.

Second screenshot, 14k timers in the bucket list. The timer subsystem ended up choking to death. The timer subsystem is not background and will stall the Master Controller, putting every other subsystem on queue and effectively soft crashing the server.

Third screenshot, eye cancer.

Fourth screenshot - Every since oranges last tweaked Glowshrooms, he has stated that if any more issues about them arose they would be deleted. This is me carrying out that ultimatum by removing the root cause of Glowshroomy related issues - Their spreading.

**Review by LemonInTheDark — APPROVED** [review]
Even if they weren't running in a timer loop this would bring any good subsystem to its knees. Let's just send it

**Comment by Kylerace** [comment]
glowshrooms are a consistent source of shitcode that chokes out the server

**Comment by Fox-McCloud** [comment]
It's still possible to spam these all over, it's worth noting---dunno how much of an issue that will be, in and of itself---a couple of bagfulls full of these things could still result in hundreds in only a couple of minutes; no more timer SS overwhelming though.

**Comment by Krysonism** [comment]
Fix the math instead...

**Comment by Pugie** [comment]
God bless you, and god bless America

**Comment by cacogen** [comment]
I feel like there must be a way to limit the spread to prevent this from happening. For example, each planted glowshroom could only spread in a radius of x tiles (I think this may already be the case though) and if x glowshroom clusters are already processing, don't allow new ones to process until they've stopped.

**Comment by Timberpoes** [comment]
I will not be compromising on this PR, so theorycraft with that in mind. It is highly unlikely you'll get glowshroom spread returning as you'll probably find it blocked by at least one maintainer.

The fact players can still spam these by manually placing them is fine. No timers. No processing. No choking the server to death.

**Comment by cacogen** [comment]
I wrote that not expecting to change anything. But for the sake of argument, would what I suggested (only process a certain number at once while leaving the rest dormant) be possible or work?

**Comment by Krysonism** [comment]
image

The code is just bugged lol. 

The children all have the same endurance as the first glowshroom planted and just creep across the station. There is also a max where there should be a min.

File: code/game/objects/effects/glowshroom.dm
Comment (kriskog) [inline code/game/objects/effects/glowshroom.dm:75]: Clamp these stats, and it fixes these issues. Every time the shrooms spread, the init proc has mutate_stats=TRUE. Since the values arent clamped, you'll get some stupid and unexpected values breaking the rest of the logic. For instance, the way the yield loop is set up it will still run on negative yield, so a yield of -30 would attempt to make 32 new mushrooms with yield of -30 with a rand(-3,2) modifier. The negative value also breaks the yield based spread termination.

**Comment by kriskog** [comment]
On this note, the timer subsystem really should have some protections against filling it up with similar or identical timers. It might break the usage of recursive timers like this though.

File: code/game/objects/effects/glowshroom.dm
Comment (Timberpoes) [reply code/game/objects/effects/glowshroom.dm:75]: I have no intent of fixing the issues **again**. Oranges said a bit of 3 months ago when he last fixed Glowshrooms that if something like this happened again they'd be removed. I am tired of the constant ping-pong game of when Glowshrooms next break the game due to their ability to generate large amounts of processing from very limited player inputs, like a DRDoS against the MC where one player can click once and spawn an entity that ends up spawning a hundred timers over its lifetime.

In addition, I don't like the Glowshroom spreading in general. They keep breaking things. But not breaking things in the "Oh, gameplay mechanic goes brrr way" - It does it in the "Master Controller goes brrrr, maintainer has to log on, diagnose and force the shift to end" kind of way. Between myself and Potato on two shifts last night we couldn't get the shift to recover from SStimer issues without disabling timers entirely, which renders the shift completely unplayable.

They're divisive. They're overused. I've spoken to @coiax about plans for something different and they're willing to implement something new ala Conway's Game of Life as a spreading mechanic, or alternate methods to spread this complete eye cancer across the station.

Comment (kriskog) [reply code/game/objects/effects/glowshroom.dm:75]: If we remove everything with bugs, there wont be anything left to play with 🙂

Comment (Pugie) [reply code/game/objects/effects/glowshroom.dm:75]: Make an exception. Just this once for something exceptionally awful :)))

**Comment by spessbro** [comment]
could we just make them spread slower? i like being able to yell at botany

**Comment by Krysonism** [comment]
> could we just make them spread slower? i like being able to yell at botany

Spreading is completely broken ever since the decay mechanic was introduced. 

There is a competing PR up to actually fix the spread once and for all.

**Comment by Pugie** [comment]
Glowshrooms are non-murdery enough to soft grief, too shit to use as antag.
The only positive glowshrooms have is that they can be used to make unstable mutagen, I'm pretty sure that is the only reason that this PR doesn't fully remove them. Just fully remove them as left 4 zed is better than mutagen now anyways.

**Comment by kriskog** [comment]
I do prefer #56981 assuming it performs well.

**Comment by Farquaar** [comment]
I hate glowshrooms, but I don't want to see them removed. I think removing their spread is more than fine. With this PR, you can still uglify the station with them, but the game won't do your dirty work for you.

**Comment by kriskog** [comment]
https://streamable.com/qswh1o is at 10x spread speed for visualizing, from the spread rework currently up. The decay means you cannot just plant a couple and it will spread everywhere.

**Comment by Mothblocks** [comment]
I prefer the rework as well, but keeping this up in case Floyd doesn't do the feedback and it goes stale.

**Comment by Mothblocks** [comment]
Superseded by Floyd's change.

**Pull request closed without merging**
